### PR TITLE
Parse and load local dapps directly from filesystem

### PR DESCRIPTION
**Pull request by axelchalon**
Local dapps are now located in `~/.config/Parity-UI/dapps/`. Local dapps served by Parity are ignored and need to be moved to this new location.

Prevent local dapps from using the appId of a non-local dapp (for which the user might already have approved permissions): `{"id": "v1"}` is inconspicuous. Local dapps ids are now prefixed with `LOCAL-` (caps to prevent accidental use of the prefix when naming one's dapp).

### CODE REVIEW

**Comment by amaurymartiny** [comment]
Nice! I'll try tomorrow.

In the meantime can you:
- check if you have npm v6. I think the 9k line diff came from me who installed npm v6 and ran npm install, and there's an updated format in package-lock. (in a separate PR we'll switch to yarn)
- `git remote add gitlab ssh://git@gitlab.parity.io/Parity-JS/shell.git` and `git push gitlab`. Automatic hooks on gitlab are broken, we need for now to manually push to gitlab too

File: src/util/host.js [hunk 1]
```diff
@@ -58,3 +57,22 @@ export function getBuildPath () {
...
+  const userData = typeof window !== 'undefined'
+    ? window.require('electron').remote.app.getPath('userData')
+    : require('electron').app.getPath('userData');
+
+  return path.join(userData, 'dapps');
+}
+
+// FIXME
+// Need to find a better way to be able to import this file
+// from electron/ as well as src/
```
Comment (amaurymartiny) [inline src/util/host.js]: I think it's cleaner to have 2 separate util/host.js, one for browser and one for electron, maybe even a better name than host for the electron one.

File: src/util/dapps.js [hunk 2]
```diff
@@ -128,16 +132,49 @@ export function fetchBuiltinApps (api) {
...
+    .then(dapps =>
+      dapps.filter(({ manifest }) => manifest))
+    .then(dapps =>
+      dapps.map(({ filename, manifest: { id, localUrl, iconUrl, ...rest } }) => (
+        {
+          ...rest,
+          type: 'local',
+          // Prevent using the appId of an existing non-local dapp with already approved permissions
+          id: `LOCAL-${id}`,
+          folderName: 'XYZ',
```
Comment (amaurymartiny) [inline src/util/dapps.js]: What does XYZ mean?

File: src/util/dapps.js [hunk 3]
```diff
@@ -128,16 +132,49 @@ export function fetchBuiltinApps (api) {
...
+          visible: true,
```
Comment (amaurymartiny) [inline src/util/dapps.js]: I think you can remove `visible: true`, I think it's not used. Anyways, this should not be how we manage visibility of dapps, there's Browse Dapps for that.

File: src/util/dapps.js [hunk 4]
```diff
@@ -128,16 +132,49 @@ export function fetchBuiltinApps (api) {
...
+          localUrl: localUrl || `file://${dappsPath}/${filename}/index.html`,
```
Comment (amaurymartiny) [inline src/util/dapps.js]: Maybe we could rename this field to `url ` like in [dappsBuiltin.json](https://github.com/parity-js/shell/blob/master/src/Dapps/dappsBuiltin.json#L130)

File: electron/index.js [hunk 5]
```diff
@@ -128,3 +140,5 @@ app.on('activate', () => {
     createWindow();
   }
 });
+
+app.setPath('userData', path.join(app.getPath('appData'), 'Parity-UI'));
```
Comment (amaurymartiny) [inline electron/index.js]: Why is this needed?

Comment (axelchalon) [reply electron/index.js]: Otherwise `userData` is `~/.config/Electron/`. Comment from the commit:

```
Electron stores the user data by default in `~/.config/<app name>`

The app name is retrieved from the package.json when Electron is
run on a folder (e.g. `electron .`); default app name 'Electron'
is used if Electron is run directly on a script, as is the case.

Electron can be run either on .build/ or on .electron/ (dev),
so it's not practical to use package.json to set the app name
(.electron/ has no package.json); hence it's better in our case
to define it in the code.
```

Setting the app name programatically doesn't seem to have an effect on the default `userData` value (it's probably computed only once at the beginning).

File: src/util/dapps.js (hunk 2 above)
Comment (axelchalon) [reply src/util/dapps.js]: Sorry, forgot to push updated commit! This line should be removed

File: src/util/dapps.js (hunk 3 above)
Comment (axelchalon) [reply src/util/dapps.js]: The local dapps don't show up by default if I remove `visible: true` here (although they still can be displayed manually using `Browse Dapps`).

This only [sets the default visibility](https://github.com/parity-js/shell/blob/2a0a39637f6d53d5112109d5e922f12def85ff3b/src/Dapps/store.js#L349-L353) of the dapp; do you think this is problematic?

File: src/util/dapps.js (hunk 4 above)
Comment (axelchalon) [reply src/util/dapps.js]: hmm, I don't think `url` in dappsBuiltin.json is used anywhere AFAIK? (are you okay with me removing this field from dappsBuiltin.json?) Their actual URL is `(path to electron build)/dapps/(id)/` so there doesn't seem to be any relationship

However we could compute the localUrl of built-in dapps when parsing them [here](https://github.com/parity-js/shell/blob/master/src/util/dapps.js#L117) for consistency (add ``app.localUrl = `file://${builtinDappsPath}/${app.id}` ``); then we can remove [this logic](https://github.com/parity-js/shell/blob/61b286d2578c4d926be7edc262f6ee5b92f4172e/src/Dapp/dapp.js#L184-L198) from `Dapp.js` and have it be simply ``src = `${app.localUrl}?appId=${app.id} `` just like local dapps (we can decide to keep the query string or ditch it for built-in dapps; if we keep it `inject.js will be simpler). What do you think?

Comment (amaurymartiny) [reply src/util/dapps.js]: Agreed with everything: compute it upstream, so that dapps.js looks simpler, and add query string for consistency.

If it can make everything simpler, then go for it.

File: electron/index.js (hunk 5 above)
Comment (amaurymartiny) [reply electron/index.js]: userData will be `~/.config/Electron/` when running electron, but will be `~/.config/parity-ui/` once builded.

I think we can keep `app.getPath('userData')` everywhere?

File: src/util/dapps.js (hunk 4 above)
Comment (amaurymartiny) [reply src/util/dapps.js]: On a related note: maybe add a README.md in a folder that described all the fields in the manifest.json. Not urgent, can be done later, eg in the wiki .

**Comment by axelchalon** [comment]
You're right; my npm version was outdated. I updated to npm v6 and now package-lock.json is only a 303-line diff :p Didn't commit it.

File: electron/index.js (hunk 5 above)
Comment (amaurymartiny) [reply electron/index.js]: Deleted my previous comment because I just understood your reasoning.

A small remark: should it be `parity-ui` lowercase? When I run the binaries on mac, they have `./path/to/parity-ui` as their userData path.

If it's indeed lowercase, maybe use `import { name } from '../package.json'` instead of hardcoding a string.

Comment (axelchalon) [reply electron/index.js]: Okay, let's switch to lowercase then! Didn't think of importing from package.json, good idea!

File: src/util/dapps.js [hunk 6]
```diff
@@ -128,16 +132,48 @@ export function fetchBuiltinApps (api) {
 }
 
 export function fetchLocalApps (api) {
```
Comment (amaurymartiny) [inline src/util/dapps.js]: We can remove this arg
